internal/attestations/authorization.go

```diff
@@ -0,0 +1,174 @@
+	if err := json.Unmarshal(predicateBytes, predicateInterface); err != nil {
+		return nil, err
+	}
+
+	predicateStruct, err := structpb.NewStruct(*predicateInterface)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ita.Statement{
+		Type: ita.StatementTypeUri,
+		Subject: []*ita.ResourceDescriptor{
+			{
+				Digest: map[string]string{digestGitCommitKey: toTargetID},
+			},
+		},
+		PredicateType: AuthorizationPredicateType,
+		Predicate:     predicateStruct,
+	}, nil
+}
+
+// SetAuthorizationAttestation writes the new authorization attestation to the
+// object store and tracks it in the current attestations state.
+func (a *Attestations) SetAuthorizationAttestation(repo *git.Repository, env *sslibdsse.Envelope, refName, fromID, toID string) error {
+	if err := validateAuthorization(env, refName, fromID, toID); err != nil {
+		return err
+	}
+
+	envBytes, err := json.Marshal(env)
+	if err != nil {
+		return err
+	}
+
+	blobID, err := gitinterface.WriteBlob(repo, envBytes)
+	if err != nil {
+		return err
+	}
+
+	if a.authorizations == nil {
+		a.authorizations = map[string]plumbing.Hash{}
+	}
+
+	a.authorizations[AuthorizationPath(refName, fromID, toID)] = blobID
+	return nil
+}
+
+// RemoveAuthorizationAttestation removes a set authorization attestation
+// entirely. The object, however, isn't removed from the object store as prior
+// states may still need it.
+func (a *Attestations) RemoveAuthorizationAttestation(refName, fromID, toID string) error {
+	authPath := AuthorizationPath(refName, fromID, toID)
+	if _, has := a.authorizations[authPath]; !has {
+		return ErrAuthorizationNotFound
+	}
+
+	delete(a.authorizations, authPath)
+	return nil
+}
+
+// GetAuthorizationAttestationFor returns the requested authorization
+// attestation (with its signatures).
+func (a *Attestations) GetAuthorizationAttestationFor(repo *git.Repository, refName, fromID, toID string) (*sslibdsse.Envelope, error) {
+	blobID, has := a.authorizations[AuthorizationPath(refName, fromID, toID)]
+	if !has {
+		return nil, ErrAuthorizationNotFound
+	}
+
+	envBytes, err := gitinterface.ReadBlob(repo, blobID)
+	if err != nil {
+		return nil, err
+	}
+
+	env := &sslibdsse.Envelope{}
+	if err := json.Unmarshal(envBytes, env); err != nil {
+		return nil, err
+	}
+
+	if err := validateAuthorization(env, refName, fromID, toID); err != nil {
+		return nil, err
+	}
+
+	return env, nil
+}
+
+// AuthorizationPath constructs the expected path on-disk for the authorization
+// attestation.
+func AuthorizationPath(refName, fromID, toID string) string {
+	return path.Join(refName, fmt.Sprintf("%s-%s", fromID, toID))
+}
+
+func validateAuthorization(env *sslibdsse.Envelope, refName, fromID, toID string) error {
+	payload, err := env.DecodeB64Payload()
+	if err != nil {
+		return err
+	}
+
+	attestation := &ita.Statement{}
+	if err := json.Unmarshal(payload, attestation); err != nil {
+		return err
+	}
+
+	if attestation.Subject[0].Digest["gitCommit"] != toID {
```

Comment:
I'd iterate through here, since there could be multiple digests.

Reply:
Happy to add that here. However, do we foresee allowing users to attach attestations generated outside of gittuf for these workflows? The reason I didn't add this is because I expected gittuf to be the sole producer.